### ISSUE TITLE
fairplay usage

### ISSUE THREAD

**Issue by life4u2c** [google/shaka-player#2557]
**Have you read the [FAQ](https://bit.ly/ShakaFAQ) and checked for duplicate open issues?**
Yes

**What version of Shaka Player are you using?**
2.5.11

**Can you reproduce the issue with our latest release version?**
Yes

**Can you reproduce the issue with the latest code from `master`?**
Yes

**Are you using the demo app or your own custom app?**
custom

**If custom app, can you reproduce the issue using our demo app?**
not entirely sure how to configure a Fairplay example in the demo app

**What browser and OS are you using?**
Safari Version 13.1 (15609.1.20.111.8)
MacOS Catalina Version 10.15.4 (19E287)

**For embedded devices (smart TVs, etc.), what model and firmware version are you using?**
N/A

**What are the manifest and license server URIs?**
manifest uri
https://content.downlynk.com/f467693928ee4d27b07e02a9a728c99e.m3u8?rmt=fps&skd=1
NOTE about manifest uri, I would want to expect it to work with skd=0 on our servers, but that doesn't seem to be making the license requests as I expect
hls certificate url
https://x-drm.downlynk.com/fairplay/public_key/06d27fd7196043a88a8454f729f79bbb.cer

**What did you do?**
I will comment with an example code on the bug below. But essentially, I am trying to setup initDataTransform method for fairplay and this method appears to not be working. Also, it seems a little out of date from the JSDocs tutorial.

**What did you expect to happen?**
I expect the initDataTransform method to be called. I have thrown in some console statements and not seeing these resulting in console logs.

**What actually happened?**
As described here https://shaka-player-demo.appspot.com/docs/api/tutorial-fairplay.html
I want to setup the initDataTransform but it doesn't seem to be working.

**Comment by life4u2c** [comment]
We are wanting to recommend shaka player but it's a bit harder to recommend without knowing how to tell our customers how to get fairplay working with it

I'm sure there are other things need tweaking for it to fully work, but here is a code example that illustrates not seeing the initDataTransform method working

```
<!doctype html>
<html>

<head>
    <meta charset="utf-8">
    <script src="shaka-player.compiled.js"></script>
    <title>Fairplay Example</title>
</head>

<h1>Fairplay Example: Shaka - v2.5.11</h1>

<body>

<video id="video"
       width="480" height="270"
       controls autoplay></video>

<script>
    (function (window) {
        if (!window.shaka) {
            throw 'shaka must be loaded';
        }
        let shaka = window.shaka;

        let manifestUri = 'https://content.downlynk.com/f467693928ee4d27b07e02a9a728c99e.m3u8?rmt=fps&skd=1';
        // let manifestUri = 'https://content.downlynk.com/f467693928ee4d27b07e02a9a728c99e.m3u8?rmt=fps&skd=0';
        let hlsCertificateUrl = 'https://x-drm.downlynk.com/fairplay/public_key/06d27fd7196043a88a8454f729f79bbb.cer';
        let hlsCertificate;  // loaded from url later

        function setup() {
            console.log('setup running');
            shaka.polyfill.installAll();
            if (!shaka.Player.isBrowserSupported()) {
                alert('Browser is not supported');
            }
            addPlayer();
        }

        function initPlayer() {
            let video = document.getElementById('video');
            window.video = video;
            let player = new shaka.Player(video);
            window.player = player;

            player.getNetworkingEngine().registerRequestFilter(requestFilter);
            player.getNetworkingEngine().registerResponseFilter(responseFilter);
            player.addEventListener('error', showErrorEventHandler);

            let playerOptions = {
                drm: {
                    fairPlayTransform: false,
                    advanced: {
                        'com.apple.fps.1_0': {
                            serverCertificate: new Uint8Array(hlsCertificate)
                        }
                    },
                    initDataTransform: initDataTransform
                }
            }
            console.log('playerOptions ', playerOptions);
            player.configure(playerOptions);

            player.load(manifestUri)
                .then(showLoadSuccess)
                .catch(showError);
        }

        function responseFilter(type, response) {
            console.log('response filter before license check');
            if (type !== shaka.net.NetworkingEngine.RequestType.LICENSE) {
                return;
            }

            console.log('response filter after license check');
        }

        function requestFilter(type, request) {
            console.log('request filter before license check');
            if (type !== shaka.net.NetworkingEngine.RequestType.LICENSE) {
                return;
            }
            console.log('request filter after license check');
        }

        function initDataTransform(initData, initDataType) {
            console.log('initData: ', initData);
            console.log('initDataType: ', initDataType);
            let skdUrl = arrayToString(initData);
            let spcUrl = skdUrl.replace('skd://', 'https://');
            spcUrl = spcUrl.substring(1, spcUrl.length);
            const cert = player.drmInfo().serverCertificate;
            return shaka.util.FairPlayUtils.initDataTransform(initData, spcUrl, cert);
        }

        function addPlayer() {
            fetch(hlsCertificateUrl)
                .then(addPlayerWithCertificate);
        }

        function addPlayerWithCertificate(response) {
            hlsCertificate = response.arrayBuffer();
            initPlayer();
        }

        function showLoadSuccess() {
            console.log('Player loaded successfully');
        }

        function showErrorEventHandler(event) {
            showError(event.detail);
        }

        function showError(error) {
            let message = 'see console';
            if (error.message && error.message.match) {
                let matches = error.message.match(/[^Shaka Error].+?(?=\()/);
                if (matches) {
                    message = matches[0].replace(/\./g, ' ').replace(/_/g, ' ');
                }
            }
            console.error('https://shaka-player-demo.appspot.com/docs/api/shaka.util.Error.html#value:' + error.code);
            console.error(error);
            alert('ERROR - ' + error.code + ': ' + message);
        }

        setup()
    })(window);

</script>
</body>

</html>

```

**Comment by TheModMaker** [comment]
You should try using the debug version to get better logging.  Are you getting any errors?  You don't set a license server, which is also required.  Also, the new `initDataTransform` should not include the `substring` call since there is no prefix anymore.

**Comment by joeyparrish** [comment]
@TheModMaker makes a good point.  The lack of a license server config seems to be the root issue here.  If you don't configure the player with a FairPlay license server URL, it will never try to use FairPlay.  The FairPlay tutorial doesn't explicitly mention a license server config, but the general configuration in the basic DRM tutorial still applies: https://shaka-player-demo.appspot.com/docs/api/tutorial-drm-config.html

Try adding something like this to your existing config:

```js
player.configure({
  drm: {
    servers: {
      // YOUR LICENSE SERVER GOES HERE:
      'com.apple.fps.1_0': 'https://foo.bar/fairplay',
    },
  },
});
```

Does this help?

**Comment by life4u2c** [comment]
That gets my callback to start being executed. I think I can try new things to get it to start working.

I must be misunderstanding something though, since HLS does not require a license server to be specified at the time of the request. 

The license url is in the manifest. And the url in the manifest will contain various URL params which are going to be unique per account.

This probably explains the behavior I was getting before too:
With `skd=1`, our license urls are coming back as `skd:// ...` and no license request was taking place
With `skd=0`, our license urls are coming back as 'https:// ...` and a license request was happening, but would fail with 6010 (ENCRYPTED_CONTENT_WITHOUT_DRM_INFO), I'm assuming because the player did not have the certificate available to it.

In any case, specifying the server license url does get my initDataTransform to start being called. That gets me past the scope of the bug I was submitting. I will continue to work through it. 

Also, any tips for getting fariplay url to work in the demo app? It would be helpful, but I know your busy.

Thank you for your time and guidance.

**Comment by life4u2c** [comment]
As far as extra logging, Yes, I was getting errors. Related to 3016 not playing or 6010 encrypted but no drm info. Also, with 2.5.11, `window.shaka.log` did not work for me. I'm assuming it's changed how you configure logging, but didn't track that down.

**Comment by joeyparrish** [comment]
> I must be misunderstanding something though, since HLS does not require a license server to be specified at the time of the request.

Shaka Player has always required a license server to be specified before you call `load()`.  In a multi-DRM world, we sometimes run into DASH content, for example, which contains signals for multiple DRM providers.  The application developer may not necessarily have license servers for all of them, though, and we need to negotiate a `MediaKeys` instance with the browser to determine if the content can be played.  We need to know what DRM systems the application supports, and the license server config fills this role.

You can always use a request filter to change the license request URI, though.  So if you don't know it in advance for some reason, just fill in a dummy value in the player config and do a substitution in the filter.  Check out the examples for URI-based authentication in our authentication tutorial, which are really just modifying the request URI: https://shaka-player-demo.appspot.com/docs/api/tutorial-license-server-auth.html

> Also, any tips for getting fariplay url to work in the demo app? It would be helpful, but I know you're busy.

FairPlay requires a certificate and (often) callbacks to manipulate init data and some kind of license request/response formatting.  There is no really good way to specify those callbacks in the demo app.  We have discussed adding some extra fields to the "custom content" form that would allow you to paste in JS code for those callbacks, but it gets messy very quickly.  Not only would the UI for that occupy a lot of visual real estate, but we'd have to find ways to surface errors from your callbacks (including syntax errors), and we'd have to `eval()` the code, which creates security implications we haven't had time to think through.

So, long story short, we don't have a great way to support FairPlay in "custom content" today.  We're open to suggestions, though!

> Also, with 2.5.11, window.shaka.log did not work for me. I'm assuming it's changed how you configure logging, but didn't track that down.

Are you using the compiled bundle?  Logging is turned off in production builds, so you'd need a debug build.  There is some relevant info in the debugging tutorial: https://shaka-player-demo.appspot.com/docs/api/tutorial-debugging.html

**Comment by life4u2c** [comment]
- Regarding shaka.log
So I do something like this in my sample app:
```
if (!window.shaka) {
    throw 'shaka-player.compiled.js must be loaded before player.js';
}
if (!window.shaka.log) {
    throw 'shaka-player.compiled.debug.js must be loaded before player.js (this allows for things like shaka.log)';
}
```
I am including it by downloading with npm and adding it in my sample app like this:
````
<script src="/node_modules/shaka-player/dist/shaka-player.compiled.debug.js"></script>
````
From 2.5.10 to 2.5.11 I start to receive the error when shaka.log is not available. But I did not trace this down nor did I try latest master yet either.

- Back to getting my app working again, if you have time for some more guidance.

I am able to trigger the initDataTransform by adding the drm.servers for FPS. I am currently hard coding the value for this particular response into the licenseServer.
However, I have issues when trying to use the initDataTransform. What is the response expected to be? I suppose what I'm really after is that I can get to spc uri so I can use that for my license request. When I use initDataTransform, my license request never takes place. I'm assuming because I'm doing something wrong in this method. 

If I don't use initDataTransform, I can hard code the drm license server by setting it as you suggested in the request filter without using the initDataTransform, but is there another way to get the spc uri besides the initDataTransform? Or to get it to work and behave like I'm not using the initDataTransform?

I am having problems parsing my spc, but that's probably something I am doing silly and will get a teammate to take a look at that part for me.

In any case, I appreciate your help.

**Comment by TheModMaker** [comment]
See #2565 for `shaka.log`.  Use the uncompiled version for now to change the logging.

The `initDataTransform` changes the init data we get from the media segment before we pass it to the browser.  The browser expects certain data in the data, one important thing is the content ID.  If you use a custom content ID, you'll need to read the init data from the media and get your version of the content ID before it is sent to the browser.  You should be able to just return this [initDataTransform](https://shaka-player-demo.appspot.com/docs/api/shaka.util.FairPlayUtils.html#.initDataTransform).

To change the license server URL, you need to use a network request filter.  Something like this:

```js
player.getNetworkingEngine().addRequestFilter((type, request) => {
  if (type == shaka.net.NetworkingEngine.RequestType.LICENSE) {
    request.uri = myUri;
  }
});
```

**Comment by joeyparrish** [comment]
Small correction:

There is no `uri` on the request.  The correct field is an array called `uris` (because we support a general URI-based fallback on failure).

```js
    request.uris = [myUri];
```

**Comment by life4u2c** [comment]
OK, thank you for your responses. 

I am already doing that change to the license uri, where we set `request.uris`. It looks like this now as a change from the above example: (there are some other minor tweaks, but here is the requestFilter now)
```
            if (type !== shaka.net.NetworkingEngine.RequestType.LICENSE) {
                return;
            }

            const originalPayload = new Uint8Array(request.body);
            const base64Payload = shaka.util.Uint8ArrayUtils.toBase64(originalPayload);
            request.headers['Content-Type'] = 'application/json';
            request.body = `{"spc": "` + base64Payload + `"}`;
            request.uris = ['https://content-dusw1.downlynk.com/fp-spc/?b=f467693928ee4d27b07e02a9a728c99e&v=f467693928ee4d27b07e02a9a728c99e&r=d&pbs=a7e327bc801a467cb55'];
```
As you can see, the request.uris is a hard coded uri for now.

However, the actual uri is in the asset request. I can see it if I use the initDataTransform, but I can't get a requestFilter to be called when using the initDataTransform method. I'm assuming I'm doing something wrong. I also can't get the spc uri in the request filter. I've tried various things in this method. I'm not trying to do special content id transform, but I want access to this uri since it is the actual license uri I want to use. We are doing things like adding query parameters and subdomains to the uri, so we are not specifying a generic license uri like we do with DASH for playready and widevine.

**What I'm after:**
A way to get the spc uri from the asset call to set it as the drm license request uri

1. The manifest call looks like this: `https://content.downlynk.com/f467693928ee4d27b07e02a9a728c99e.m3u8?rmt=fps&skd=1`
2. Then, the asset call looks something like this `https://content-ause1-dp-2.downlynk.com/f467693928ee4d27b07e02a9a728c99e/d.m3u8?rmt=fps&skd=1&pbs=a8c90527a3a84f62b033db6cf6ef2792`
3. Finally, the asset call has the key method information including the license uri `#EXT-X-KEY:METHOD=SAMPLE-AES,URI="skd://content-ause1-dp-2.downlynk.com/fp-spc/?b=f467693928ee4d27b07e02a9a728c99e&v=f467693928ee4d27b07e02a9a728c99e&r=d&pbs=a8c90527a3a84f62b033db6cf6ef2792&sig=34c7143ec111a722d500d6cf21443cbee9d3547efea86e18efa6bf7d1c913dec",KEYFORMAT="com.apple.streamingkeydelivery",KEYFORMATVERSIONS="1"`

**Why I was looking at the initDataTransform**
I was thinking I want to do the transform here from `skd:// ...` to https:// ...` This way the license request uri would be correct. I also can accomplish this on my server by making `skd=0` during the manifest request, but then the request filter never gets called.

**Why I need the request filter**
We are doing json for the spc to our server. This is special to our servers, although we could look at allowing a different format on our servers here, we are only doing json at the moment.

I know it takes time to even just read these comments, but I am grateful for your help.

**Comment by joeyparrish** [comment]
You can use `initDataTransform` to access/transform the init data and compute the license server URI you want.  You can store that URI somewhere during the `initDataTransform` callback, then use in it the request filter.

Does that make sense?

**Comment by life4u2c** [comment]
Yes, that makes sense. I've already done something like this. However, whenever I get my `initDataTransform` method configured, my filter never gets called. I'm doing something like in the example below. I don't necessarily want the `initDataTransform` to do anything. I have also tried returning just an empty array. But in the things I have tried, the `requestFilter` never gets called.

```
// Request filter is setup like this when configuring the player
// player.getNetworkingEngine().registerRequestFilter(requestFilter);

let licenseUri;

function initDataTransform(initData, drmInfo) {
    let skdUrl = shaka.util.StringUtils.fromBytesAutoDetect(initData);
    licenseUri = skdUrl.replace('skd://', 'https://');
    const cert = drmInfo.serverCertificate;
    return shaka.util.FairPlayUtils.initDataTransform(initData, licenseUri, cert);
}

function requestFilter(type, request) {
    if (type !== shaka.net.NetworkingEngine.RequestType.LICENSE) {
        return;
    }

    const originalPayload = new Uint8Array(request.body);
    const base64Payload = shaka.util.Uint8ArrayUtils.toBase64(originalPayload);
    request.headers['Content-Type'] = 'application/json';
    request.body = '{"spc": "' + base64Payload + '"}';
    request.uris = [licenseUri];
}
```

**Comment by joeyparrish** [comment]
Is your request filter registered?

**Comment by life4u2c** [comment]
Yes, my request filter is registered. If I configure the `initDataTransform`, I can't seem to get the filter to be called. If I take the `initDataTransform` configuration out of the player options, my request filter is then called.

**Comment by joeyparrish** [comment]
Are you listening for error events from the player object?  Are you checking for a rejected Promise from player.load()?  Either of these could indicate an error that stopped the load process before the license request could occur.

**Comment by life4u2c** [comment]
Yes, I am catching those errors. I am getting the error below. I'm assuming because I am doing something wrong in the `initDataTransform` method.

FAILED_TO_GENERATE_LICENSE_REQUEST
https://shaka-player-demo.appspot.com/docs/api/shaka.util.Error.html#value:6006
<img width="602" alt="image" src="https://user-images.githubusercontent.com/6334897/81842595-e3612a80-9511-11ea-9799-7dc7c6b1ead4.png">

**Comment by joeyparrish** [comment]
Here's the flow:

1. Get init data from the browser
2. Transform the init data
3. Pass the init data to the CDM to generate a license request
4. Filter that license request
5. Send the license request to the license server
6. Get back a license response
7. Filter that response
8. Pass the response to the CDM to consume the license

You're getting an error at step 3, so your filter can't be called in step 4.

**Comment by TheModMaker** [comment]
My guess would be the browser is rejecting your content ID.  Try storing the URL and using the default content ID calculation:


```js
function initDataTransform(initData, drmInfo) {
    let skdUrl = shaka.util.StringUtils.fromBytesAutoDetect(initData);
    licenseUri = skdUrl.replace('skd://', 'https://');
    const cert = drmInfo.serverCertificate;
    const contentId = shaka.util.FairPlayUtils.defaultGetContentId(initData);
    return shaka.util.FairPlayUtils.initDataTransform(initData, contentId, cert);
}
```

**Comment by life4u2c** [comment]
👍 That allowed my filter request to be called 👍 
I still have to get my spc correct but I don't think that is for your player. I did try to use a default content id at some point, but that would have been before I added the server license uri in the player option with the 'fake' uri. I forgot to give that a try again after.

I think, I am passed my blockers for my sample app with shaka player. I still something I have to figure out and see it get passed those last few steps in the flow, but I don't think I'm blocked.

I really appreciate your help. Thank you for your time and timely responses.

**Comment by joeyparrish** [comment]
We're glad we were able to help.  Thanks for your patience with us!
